dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/DialogueRoutePlanner.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.hc5;
+
+import java.net.InetAddress;
+import java.net.ProxySelector;
+import javax.annotation.Nullable;
+import org.apache.hc.client5.http.HttpRoute;
+import org.apache.hc.client5.http.impl.routing.SystemDefaultRoutePlanner;
+import org.apache.hc.client5.http.routing.HttpRoutePlanner;
+import org.apache.hc.core5.http.HttpException;
+import org.apache.hc.core5.http.HttpHost;
+import org.apache.hc.core5.http.protocol.HttpContext;
+
+/**
+ * This implementation wraps the default route planner, but allows a specific pre-resolved address to be used.
+ * Client instances are shared between all URIs, so this route planner cannot know resolved addresses at the
+ * moment it's created, and instead must rely on callers passing the resolved address using the
+ * {@link HttpContext}.
+ */
+final class DialogueRoutePlanner implements HttpRoutePlanner {
+    private static final String ATTRIBUTE = "dialogueResolvedAddress";
+    private final HttpRoutePlanner delegate;
+
+    DialogueRoutePlanner(ProxySelector proxySelector) {
+        delegate = new SystemDefaultRoutePlanner(proxySelector);
+    }
+
+    @Override
+    public HttpRoute determineRoute(HttpHost host, HttpContext context) throws HttpException {
+        HttpRoute route = delegate.determineRoute(host, context);
+        if (route.getTargetHost().getAddress() == null) {
```

Comment:
when will this actually be true?

Reply:
This is always true in the cases I've observed, it's a defensive check to make sure we're not clobbering the target address that has already been bound.

Reply:
oh that's interesting - so name resolution is currently happening further down (or perhaps not at all in hc5) and this is just a convenient way of setting a resolved address ahead of time?

Reply:
Yep -- this sets the address ahead of time, which tells hc5 that it doesn't need to interact with the dns resolver at all. That happens here: https://github.com/apache/httpcomponents-client/blob/661589e386e204fbdba1fbd7dac4f90ca5222c7f/httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/DefaultHttpClientConnectionOperator.java#L169-L176